internal/cloudsql/instance.go

```diff
@@ -0,0 +1,209 @@
+	return fmt.Sprintf("%s:%s:%s", c.project, c.region, c.name)
+}
+
+// parseConnName initializes a new connName struct.
+func parseConnName(cn string) (connName, error) {
+	b := []byte(cn)
+	m := connNameRegex.FindSubmatch(b)
+	if m == nil {
+		return connName{}, fmt.Errorf("invalid instance connection name - expected PROJECT:REGION:ID")
+	}
+
+	c := connName{
+		project: string(m[1]),
+		region:  string(m[3]),
+		name:    string(m[4]),
+	}
+	return c, nil
+}
+
+// refreshResult is a pending result of a refresh operation of data used to connect securely. It should
+// only be initialized by the Instance struct as part of a refresh cycle.
+type refreshResult struct {
+	md     metadata
+	tlsCfg *tls.Config
+	err    error
+
+	// timer that triggers refresh, can be used to cancel.
+	timer *time.Timer
+	// indicates the struct is ready to read from
+	ready chan struct{}
+}
+
+// Cancel prevents the instanceInfo from starting, if it hasn't already started. Returns true if timer
+// was stopped successfully, or false if it has already started.
+func (i *refreshResult) Cancel() bool {
+	return i.timer.Stop()
+}
+
+// Wait blocks until the refreshResult attempt is completed.
+func (i *refreshResult) Wait(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-i.ready:
+		return i.err
+	}
+}
+
+// Instance manages the information used to connect to the Cloud SQL instance by periodically calling
+// the Cloud SQL Admin API. It automatically refreshes the required information approximately 5 minutes
+// before the previous certificate expires (every 55 minutes).
+type Instance struct {
+	connName
+	client *sqladmin.Service
+	key    *rsa.PrivateKey
+
+	resultGuard sync.RWMutex
+	cur         *refreshResult
+	next        *refreshResult
+
+	// TODO: add a way to close
+}
+
+// NewInstance initializes a new Instance given an instance connection name
+func NewInstance(instance string, client *sqladmin.Service, key *rsa.PrivateKey) (*Instance, error) {
+	cn, err := parseConnName(instance)
+	if err != nil {
+		return nil, err
+	}
+	i := &Instance{
+		connName: cn,
+		client:   client,
+		key:      key,
+		cur:      nil,
+		next:     nil,
+	}
+	// For the inital refresh operation, set cur = next so that connection requests block
+	// until the first refresh is complete.
+	i.resultGuard.Lock()
+	i.cur = i.scheduleRefresh(0)
+	i.next = i.cur
+	i.resultGuard.Unlock()
+	return i, nil
+}
+
+// Connect returns a secure, authorized net.Conn to a Cloud SQL instance.
+func (i *Instance) Connect(ctx context.Context) (net.Conn, error) {
+	i.resultGuard.RLock()
+	res := i.cur
+	i.resultGuard.RUnlock()
+	err := res.Wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Add better ipType support, including an opt to specify.
+	addr := net.JoinHostPort(res.md.ipAddrs["PUBLIC"], "3307")
+	conn, err := proxy.Dial(ctx, "tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	if c, ok := conn.(*net.TCPConn); ok {
```

Comment:
I suppose `conn` could also be a `*net.UnixConn`. Is that correct?

Reply:
No - this is the outgoing connection to a Cloud SQL instance, so it'll probably always be a TCP connection.

Reply:
In that case, I think this should be:

``` go
c, ok := conn.(*net.TCPConn)
if !ok {
  return nil, errors.New("dailed to cast connection to TCPConn")
}

// ... and then continue
```

Reply:
I understand the desire to do this, but not sure if we should here. While I'm fairly certain we'll always have a TCP connection, I'm less confident that `proxy.Dial` will always manifest that as a `net.TCPConn`. Since `proxy.Dial` is only guaranteed to return a `net.Conn`, I'd rather not error out just because it was an unexpected interface. 

I can add some logging if you think it's preferable over just silently skipping these operations, but it doesn't feel like a good idea to terminate early.

Reply:
I see. So this is more of a best effort of setting keep alive then. That seems reasonable then.